catalog/dags/data_refresh/distributed_reindex.py

```diff
@@ -0,0 +1,425 @@
+    """
+    if environment != PRODUCTION:
+        raise AirflowSkipException("Skipping instance creation in local environment.")
+
+    ec2_hook = EC2Hook(aws_conn_id=aws_conn_id, api_type="client_type")
+    instances = ec2_hook.conn.run_instances(
+        MinCount=1,
+        MaxCount=1,
+        LaunchTemplate={
+            "LaunchTemplateName": INDEXER_LAUNCH_TEMPLATES.get(target_environment),
+            "Version": str(launch_template_version),
+        },
+        # Name the instance by applying a tag
+        TagSpecifications=[
+            {
+                "ResourceType": "instance",
+                "Tags": [
+                    {
+                        "Key": "Name",
+                        "Value": f"catalog-indexer-worker-{target_environment}",
+                    },
+                ],
+            },
+        ],
+    )["Instances"]
+
+    if not len(instances) == 1:
+        raise Exception(
+            f"Expected one new instance, but {len(instances)} were created."
+        )
+
+    return instances[0]["InstanceId"]
+
+
+@task.sensor(poke_interval=60, timeout=3600, mode="reschedule")
+def wait_for_worker(
+    environment: str,
+    instance_id: str,
+    aws_conn_id: str = AWS_CONN_ID,
+):
+    """Await the EC2 instance with the given id to be in a healthy running state."""
+    if environment != PRODUCTION:
+        raise AirflowSkipException("Skipping instance creation in local environment.")
+
+    ec2_hook = EC2Hook(aws_conn_id=aws_conn_id, api_type="client_type")
+    result = ec2_hook.conn.describe_instance_status(InstanceIds=[instance_id])
+
+    instance_status = result.get("InstanceStatuses", [])[0]
+    state = instance_status.get("InstanceState", {}).get("Name")
+    status = next(
+        status.get("Status")
+        for status in instance_status.get("InstanceStatus", {}).get("Details", [])
+        if status.get("Name") == "reachability"
+    )
+
+    return PokeReturnValue(
+        # Sensor completes only when the instance is running and has finished initializing
+        is_done=(state == "running" and status == "ok")
+    )
+
+
+@task(
+    # Run locally when create instance tasks are skipped
+    trigger_rule=TriggerRule.NONE_FAILED
+)
+def get_instance_ip_address(
+    environment: str, instance_id: str, aws_conn_id: str = AWS_CONN_ID
+):
+    if environment != PRODUCTION:
+        return "catalog_indexer_worker"
+
+    ec2_hook = EC2Hook(aws_conn_id=aws_conn_id, api_type="client_type")
+    reservations = ec2_hook.describe_instances(instance_ids=[instance_id]).get(
+        "Reservations"
+    )
+
+    # `Reservations` is a list of dicts, grouping instances by the launch request that
+    # started them. Because we are querying a single InstanceId we expect to have one
+    # Reservation.
+    if not (len(reservations) == 1 and len(reservations.get("Instances", {})) == 1):
+        raise Exception("Unable to describe worker instance.")
```

Comment:
Here again, boto3 should have you covered. Here's an example of calling `describe_instances` with an instance ID that doesn't exist:

```
>>> ec2_client.describe_instances(InstanceIds=["i-0de0dc5cd17fc5f2c"])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/sara/projects/openverse-infrastructure/.venv/lib/python3.12/site-packages/botocore/client.py", line 565, in _api_call
    return self._make_api_call(operation_name, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/sara/projects/openverse-infrastructure/.venv/lib/python3.12/site-packages/botocore/client.py", line 1021, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.exceptions.ClientError: An error occurred (InvalidInstanceID.Malformed) when calling the DescribeInstances operation: Invalid id: "i-0de0dc5cd17fc5f2c"
```

Reply:
My understanding was that there _can_ be more than one reservation, which is what I was trying to catch -- but actually I'm only passing one instance ID so that isn't necessary either! I think this was necessary in the previous design where I was describing multiple instances at once 😅 Awesome, good to get rid of some potentially confusing checks.